app/auth/[...auth]/route.ts

```diff
@@ -14,6 +17,38 @@ export const { GET, POST } = auth.createAuthRouteHandlers({
       redirect("/error?error=email-verification-required");
     }
 
+    const url = new URL(req.url);
+    const code = url.searchParams.get("code");
+
+    if (!code) {
+      const searchParamsError = url.searchParams.get("error");
+      console.error("Code not found: ", searchParamsError, error);
+      redirect("/error?error=code-not-found");
+    }
+
+    const verifierCookie = req.cookies.get("edgedb-pkce-verifier");
+    if (!verifierCookie) {
+      console.error("Could not find 'verifier' in the cookie store. Is this the same user agent/browser that started the authorization flow?");
+      redirect("/error?error=verifier-not-found");
+    }
+
+    const codeExchangeUrl = new URL("token", EDGEDB_AUTH_BASE_URL);
+    codeExchangeUrl.searchParams.set("code", code);
+    codeExchangeUrl.searchParams.set("verifier", verifierCookie.value);
+
+    const codeExchangeResponse = await fetch(codeExchangeUrl.href, {
+      method: "GET",
+    });
+
+    if (codeExchangeResponse.status !== 200) {
+      console.error("Code exchange failed: ", codeExchangeResponse.status, codeExchangeResponse.statusText);
+      redirect("/error?error=code-exchange-failed");
+    }
+
+    const { auth_token, provider_token } = await codeExchangeResponse.json();
+
+    const profile = await getUserProfile(provider_token);
```

Comment:
The `provider_token` is already (potentially) in the `tokenData`, so this is all you need:

```suggestion
    const { provider_token } = tokenData;
    const profile = provider_token && await getUserProfile(provider_token);
```